ddollar/foreman
# Migrate CI from Travis to GitHub Actions

Since June 15th, 2021, the building on travis-ci.org is ceased.
This PR migrates CI from Travis CI to GitHub Actions.
Here are the details:

- Add Ruby 2.7, 3.0, 3.1 and head to the CI matrix
- Remove and ignore `Gemfile.lock`
  - Existence of `Gemfile.lock` prevents testing on multiple Ruby versions.
- Update RubyGems to the latest
  - Using old RubyGems, the following warning is displayed and the CI fails.
  ```
  1) Foreman::CLI run can run a command
     Failure/Error: expect(forked_foreman("run echo 1")).to eq("1\n")

       expected: "1\n"
            got: "Your RubyGems version (2.7.6.3) has a bug that prevents `required_ruby_version` from working for Bun...yGems to avoid future breakage and silence this warning by running `gem update --system 3.2.3`\n1\n"
  ```
- Use the latest fakefs
  - fakefs v0.10.x does not work with the latest Ruby.
- Replace `File.exists?` with `File.exist?`
  - `File.exists?` is being removed in Ruby 3.2.
- Pass a block instead of a lambda to Rspec's `expect` to suppress the following deprecation warning
  ```
  The implicit block expectation syntax is deprecated, you should pass a block rather than an argument to `expect` to use the provided block expectation matcher or the matcher must implement `supports_value_expectations?`. e.g  `expect { value }.to raise SystemExit` not `expect(value).to raise SystemExit`
  ```
- Add GitHub token permissions for workflow
  - GitHub asks developers to define workflow permissions,
    see below for securing GitHub workflows against supply-chain attacks.
    - https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/
    - https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token


`codeclimate-test-reporter` has been deprecated and the following message displayed:
```
  Code Climate's codeclimate-test-reporter gem has been deprecated in favor of
  our language-agnostic unified test reporter. The new test reporter is faster,
  distributed as a static binary so dependency conflicts never occur, and
  supports parallelized CI builds & multi-language CI configurations.

  Please visit https://docs.codeclimate.com/v1.0/docs/configuring-test-coverage
  for help setting up your CI process with our new test reporter.
```
I wanted to deal with this, but since I don't have the token of CodeClimate, I passed on it.
